File: java/src/test/java/org/training/speedrun/stage5/Stage5Test.java
package org.training.speedrun.stage5;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class Stage5Test {

    private static Stream<Arguments> generateCart() {
        return Stream.of(
                Arguments.of(new double[]{}, false, false, 0),
                Arguments.of(new double[]{2, 3}, false, false, 5),
                Arguments.of(new double[]{10, 10}, true, true, 25),
                Arguments.of(new double[]{10, 10}, true, false, 21),
                Arguments.of(new double[]{10, 10}, false, true, 24),
                Arguments.of(new double[]{1, 1, 1, 1, 1}, false, false, 5.2),
                Arguments.of(new double[]{1, 1, 1, 1, 1}, true, false, 5.45),
                Arguments.of(new double[]{1, 1, 1, 1, 1}, false, true, 6.2),
                Arguments.of(new double[]{1, 1, 1, 1, 1}, true, true, 6.45),
                Arguments.of(new double[]{2, 2, 2, 2, 2, 2, 2, 2, 2, 2}, false, false, 22),
                Arguments.of(new double[]{2, 2, 2, 2, 2, 2, 2, 2, 2, 2}, true, false, 23),
                Arguments.of(new double[]{2, 2, 2, 2, 2, 2, 2, 2, 2, 2}, false, true, 26),
                Arguments.of(new double[]{2, 2, 2, 2, 2, 2, 2, 2, 2, 2}, true, true, 27)
        );
    }

    @ParameterizedTest
    @MethodSource("generateCart")
    void compute_cart_price(double[] c, boolean v, boolean b, double expectedPrice) {
        Assertions.assertThat(new Stage5().compute(c, v, b)).isEqualTo(expectedPrice);
    }
}

File: kotlin/src/test/kotlin/org/training/speedrun/stage5/Stage5Test.kt
package org.training.speedrun.stage5

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

internal class Stage5Test {

    private val stage5 = Stage5()

    @Test
    fun long_words_are_eligible_if_they_have_4_vowels() {
        assertThat(stage5.isEligible("azertyuiopqsdfg")).isTrue()
    }

    @Test
    fun long_words_are_eligible_if_they_have_at_least_3_special_letters() {
        assertThat(stage5.isEligible("rzmztypopqsdfg")).isTrue()
    }

    @Test
    fun short_words_are_eligible_if_they_have_at_least_5_letters_in_upper_case() {
        assertThat(stage5.isEligible("PlKZMZ")).isTrue()
    }

    @Test
    fun short_words_are_eligible_if_they_have_3_vowels() {
        assertThat(stage5.isEligible("aeio")).isTrue()
    }

    @ParameterizedTest
    @ValueSource(strings = ["pazof", "AOJGKFLDLFMG", "", "oplm", "lkoirp"])
    fun other_words_are_not_elligible(word: String) {
        assertThat(stage5.isEligible(word)).isFalse()
    }
}
